File: overlay/tests/test_touch.py
"""
test_touch.py - Test touch and gat (get and touch) commands.
"""

import os
import sys
import time
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestTouch(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_touch_extends_ttl(self):
        # Set with 2-second TTL
        self.assertEqual(self.mc.set("touch_key", "val", exptime=2), "STORED")

        # Touch to extend to 10 seconds
        self.assertEqual(self.mc.touch("touch_key", 10), "TOUCHED")

        # Sleep 2.5s (longer than original TTL)
        time.sleep(2.5)

        # Key should still exist
        res = self.mc.get("touch_key")
        self.assertIn("touch_key", res)
        self.assertEqual(res["touch_key"][0], "val")

    def test_touch_nonexistent(self):
        self.assertEqual(self.mc.touch("nonexistent_touch_key", 10), "NOT_FOUND")


if __name__ == "__main__":
    unittest.main()

File: overlay/tests/test_touch.c
/*
 * test_touch.c - Test touch commands.
 */

#include "c_harness.h"

static int test_touch_extends_ttl(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "touch_key", "val", 3, 0, 2, resp, sizeof(resp)) ==
              0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_touch(&c, "touch_key", 10, resp, sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "TOUCHED");

  test_sleep_ms(2500);

  TEST_ASSERT(client_get(&c, "touch_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "val");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_touch_nonexistent(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(
      client_touch(&c, "nonexistent_touch_key", 10, resp, sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "NOT_FOUND");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_touch_run(void) {
  RUN_TEST(test_touch_extends_ttl);
  RUN_TEST(test_touch_nonexistent);
  return 0;
}
